Stellar-ledger-entries.x

```diff
@@ -602,11 +641,27 @@ case CONTRACT_DATA:
     {
         Hash contractID;
         SCVal key;
+        ContractDataType type;
```

Comment:
This basically makes it possible to have 3 instances of the same key with different lifetimes. Is that expect?

Reply:
Yeah, we need each entry type to be in a separate key space for security purposes. Because `UNIQUE` entries have a strong, proof based guarantee of uniqueness, we need to protect against the case where a `UNIQUE` entry gets archived and a `RESOTRABLE` or `TEMPORARY` entry with the same key overwrites the `UNIQUE` entry because they can be created without a proof. So in this case, you can have 3 entries with the same key exist at the same time so long as they have different types.

Reply:
This is interesting. It would mean that the contract will have to specify the entry lifetime for every storage operation (including reads).  I don't think there are security concerns - contract controls its storage and thus shadowing a unique entry would just be a bug.

I'm actually not sure what is more of a footgun: being able to shadow a unique entry or being able to accidentally create multiple entries. The latter actually seems somewhat worse to me. Ideally we'd need to provide lifetime only when creating new entries and disallow creating/modifying the entries with inconsistent lifetimes.

Reply:
I'm not sure what you mean by lifetime here. Do you mean the entry type, like `UNIQUE`, `TEMPORARY`, etc? Assuming so, the issue is we can't consistently disallow creating and modifying entries with inconsistent types. For example, say a buggy contract calls `set(0x123, UNIQUE)` and `set(0x123, TEMP)`. Sometimes this would be disallowed if the original `UNIQUE` entry was still in the Ledger, but if it had expired, then there would be no way of disallowing the creation of the temp entry. At least in my mind `UNIQUE` entries have very strong security invariants, and the only way to maintain those consistently is via a seperate key space.

This may be a bad interface though having to specify the key and type for every operation. This made more sense originally when I imagined that the SDK would define a seperate storage class for each type, i.e. `TempStorage::set`, `UniqueStorage::set`, etc, but this may be too confusing or convoluted.

Reply:
Yeah, I meant entry type.  Security guarantees are really the contract responsibility. It's not immediately obvious if the bugs caused by creating two entries with 'almost' the same key (besides type) are better or worse than bugs caused by re-creating entry with a different type. I guess if we have 3 storage kinds, the UX won't be as confusing, but it still seems pretty clunky. I guess I'd like to see more opinions on this.